alerts/config.go

```diff
@@ -0,0 +1,129 @@
+package alerts
+
+// Data models for care team alerts.
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+// Config models a user's desired alerts.
+type Config struct {
+	Key string `json:"key" bson:"_id"`
+	// OwnerID links the user that owns and controls these alerts, i.e. the
+	// care team member.
+	OwnerID string `json:"ownerID"`
+	// InvitorID links the user whose data is shared, and will trigger alerts.
+	InvitorID string `json:"invitorID"`
+
+	UrgentLow       *WithThreshold `json:"urgentLow,omitempty"`
+	Low             *Deluxe        `json:"low,omitempty"`
+	High            *Deluxe        `json:"high,omitempty"`
+	NotLooping      *WithDelay     `json:"notLooping,omitempty"`
+	NoCommunication *WithDelay     `json:"noCommunication,omitempty"`
+}
+
+// Base describes the minimum specifics of a desired alert.
+type Base struct {
+	// Enabled controls whether notifications should be sent for this alert.
+	Enabled bool
+	// Repeat is measured in minutes.
+	Repeat DurationMinutes `json:"repeat"`
+}
+
+// DelayMixin adds a configurable delay.
+type DelayMixin struct {
+	// Delay is measured in minutes.
+	Delay DurationMinutes `json:"delay,omitempty"`
+}
+
+// ThresholdMixin adds a configurable threshold.
+type ThresholdMixin struct {
+	// Threshold is compared the current value to determine if an alert should
+	// be triggered.
+	Threshold `json:"threshold"`
+	// UnmarshalJSON prevents the use of the embedded Threshold.UnmarshalJSON.
+	UnmarshalJSON struct{}
+}
+
+// WithThreshold extends Base with ThresholdMixin.
+type WithThreshold struct {
+	Base
+	ThresholdMixin
+}
+
+// WithDelay extends Base with DelayMixin.
+type WithDelay struct {
+	Base
+	DelayMixin
+}
+
+// Deluxe extends Base with both DelayMixin and ThresholdMixin.
+type Deluxe struct {
+	Base
+	DelayMixin
+	ThresholdMixin
+}
+
+// DurationMinutes reads a JSON integer and converts it to a time.Duration.
+//
+// Values are specified in minutes.
+type DurationMinutes time.Duration
+
+func (m *DurationMinutes) UnmarshalJSON(b []byte) error {
+	if bytes.Equal(b, []byte("null")) || len(b) == 0 {
+		*m = DurationMinutes(0)
+		return nil
+	}
+	d, err := time.ParseDuration(string(b) + "m")
+	if err != nil {
+		return err
+	}
+	*m = DurationMinutes(d)
+	return nil
+}
+
+func (m DurationMinutes) Duration() time.Duration {
+	return time.Duration(m)
+}
+
+// ValueWithUnits binds a value to its units.
+//
+// Other types can extend it to parse and validate the Units.
+type ValueWithUnits struct {
```

Comment:
You may want to reuse https://github.com/tidepool-org/platform/blob/master/data/types/blood/glucose/glucose.go instead

Reply:
Thanks! I never would have seen these. :)

Hmm, I'll need to look further. I might be able to re-use some of the validation, but the actual types themselves are very different. Using Glucose would include [types.Base](https://github.com/tidepool-org/platform/blob/bdd98e0d295604a1504c6aaf6e9f2a2b8083f9d2/data/types/base.go#L33) and that's a whole lot of stuff I don't need/want.

Plus, in Glucose, both Units and Value are optional fields. Here, they're not optional, which means I can use even less of the validation. I'll take a look though. No sense in having validations in two places if it can be avoided.

Reply:
`glucose.Glucose` embeds `blood.Blood`, which embeds `types.Base`, and that's a _whole_ lot of stuff I'm definitely not interested in sharing.

Additionally, `Units` and `Value` are optional in `types.Base`, but they aren't optional here. I'm thinking I can re-use the `structure.Validator` though.